src/Compilers/Core/CodeAnalysisTest/InternalUtilities/StreamExtensionsTests.cs

```diff
@@ -0,0 +1,108 @@
+﻿
+using Microsoft.CodeAnalysis;
+using Roslyn.Test.Utilities;
+using System;
+using System.IO;
+using Xunit;
+
+namespace Roslyn.Utilities.UnitTests.InternalUtilities
+{
+    public class StreamExtensionsTests
+    {
+        [Fact]
+        public void CallsReadMultipleTimes()
+        {
+            var firstRead = true;
+            var sourceArray = new byte[] { 1, 2, 3, 4 };
+            int sourceOffset = 0;
+
+            var stream = new TestStream(readFunc: (buf, offset, count) =>
+            {
+                if (firstRead)
+                {
+                    count = count / 2;
+                    firstRead = false;
+                }
+                Array.Copy(sourceArray, sourceOffset, buf, offset, count);
+                sourceOffset += count;
+                return count;
+            });
+
+            var destArray = new byte[4];
+            var destCopy = destArray.AsImmutable();
+            // Note: Buffer is in undefined state after receiving an exception
+            Assert.Equal(sourceArray.Length, stream.TryReadAll(destArray, 0, sourceArray.Length));
+            Assert.Equal(sourceArray, destArray);
+        }
+
+        [Fact]
+        public void ExceptionsPropagate()
+        {
+            var buffer = new byte[10];
+
+            var stream = new TestStream(readFunc: (_1, _2, _3) => { throw new IOException(); });
+            Assert.Throws<IOException>(() => stream.TryReadAll(null, 0, 1));
+
+            stream = new TestStream(readFunc: (buf, offset, count) =>
+            {
+                if (offset + count > buf.Length)
+                {
+                    throw new ArgumentException();
+                }
+                return 0;
+            });
+            Assert.Equal(0, stream.TryReadAll(buffer, 0, 1));
```

Comment:
Consider adding a test where TryReadAll doesn't return 0 but reads less bytes than requested. For example, first Read succeeds but reads reads less bytes than requested and second Read returns 0.